core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/provision/service/ClusterProxyService.java

```diff
@@ -135,10 +137,26 @@ private ConfigRegistrationRequest createProxyConfigReRegisterRequest(Stack stack
     }
 
     private List<ClusterServiceConfig> serviceConfigs(Stack stack) {
-        String internalAdminUrl = internalAdminUrl(stack, ServiceFamilies.GATEWAY.getDefaultPort());
-        return asList(
-                cmServiceConfig(stack, null, "cloudera-manager", clusterManagerUrl(stack)),
-                cmServiceConfig(stack, clientCertificates(stack), CB_INTERNAL, internalAdminUrl));
+        String primaryGWInternalAdminUrl = internalAdminUrl(stack, ServiceFamilies.GATEWAY.getDefaultPort());
+        LOGGER.info("Primary GW internal admin URL is: {}", primaryGWInternalAdminUrl);
+        List<ClusterServiceConfig> clusterServiceConfigs = getClusterServiceConfigsForGWs(stack);
+        clusterServiceConfigs.add(cmServiceConfig(stack, null, "cloudera-manager", clusterManagerUrl(stack)));
+        clusterServiceConfigs.add(cmServiceConfig(stack, clientCertificates(stack), CB_INTERNAL, primaryGWInternalAdminUrl));
+        LOGGER.info("Service configs: {}", clusterServiceConfigs);
+        return clusterServiceConfigs;
+    }
+
+    private List<ClusterServiceConfig> getClusterServiceConfigsForGWs(Stack stack) {
+        List<ClusterServiceConfig> clusterServiceConfigs = new ArrayList<>();
+        List<InstanceMetaData> gatewayInstanceMetadatas = stack.getGatewayInstanceMetadata();
+        for (InstanceMetaData gatewayInstanceMetadata : gatewayInstanceMetadatas) {
+            String gatewayIp = gatewayInstanceMetadata.getPublicIpWrapper();
+            String internalAdminUrl = String.format("https://%s:%d", gatewayIp, ServiceFamilies.GATEWAY.getDefaultPort());
+            ClusterServiceConfig gatewayClusterServiceConfig = cmServiceConfig(stack, clientCertificates(stack),
+                    CB_INTERNAL + "-" + gatewayInstanceMetadata.getInstanceId(), internalAdminUrl);
```

Comment:
I guess this would be different from freeipa solution, where we use the fqdn. Is this because we don't have it at this time?

Reply:
Yes, unfortunately we don't have FQDN at this time :( My first implementation used FQDN but it failed. If somebody moves the FQDN generation part to the beginning of the stack creation then we can rewrite it, but I didn't want to rewrite that part also.